### PR TITLE
introduce [Client.disconnected]

### PR DESCRIPTION
**Pull request by cwong-ocaml**
Reopening of #4627 (I couldn't actually convince github to re-open that one for some reason, so I'm opening a new one).

I think this is necessary after all. We still don't have a good story for long-lived client sessions, and callback-based subscriptions will definitely want some mechanism to stay alive "until the last value is sent".

Signed-off-by: Cameron Wong <cwong@janestreet.com>

### CODE REVIEW

File: otherlibs/dune-rpc/private/dune_rpc_private.ml [hunk 1]
```diff
@@ -592,13 +594,24 @@ module Client = struct
...
+          | None ->
+            let+ () = Fiber.Ivar.fill on_close () in
+            None
+          | _ -> Fiber.return result
+        in
+        { read
         ; write = (fun s -> Chan.write c s)
         ; closed_read = false
         ; closed_write = false
+        ; on_close
```
Comment (rgrinberg) [inline otherlibs/dune-rpc/private/dune_rpc_private.ml:614]: Could your renamed this to `disconnected`? Or alternatively, rename `disconnected` to `on_close`? I don't see the need to make use of both terms in the code base.

**Review by rgrinberg — APPROVED** [review]
Looks good, but we should tweak the names for consistency.

File: otherlibs/dune-rpc/private/dune_rpc_private.ml (hunk 1 above)
Comment (cwong-ocaml) [reply otherlibs/dune-rpc/private/dune_rpc_private.ml]: Done

**Comment by cwong-ocaml** [comment]
I'm pretty sure the test failure is a CI issue

**Pull request merged**